receiver/aerospikereceiver/scraper.go

```diff
@@ -64,86 +64,80 @@ func newAerospikeReceiver(params component.ReceiverCreateSettings, cfg *Config,
+		client = nil
+		r.logger.Sugar().Warnf("initial client creation failed: %w", err)
 	}
-	r.emitNode(info, now, errs)
-	r.scrapeNamespaces(info, client, now, errs)
 
-	if r.config.CollectClusterMetrics {
-		r.logger.Debug("Collecting peer nodes")
-		for _, n := range strings.Split(info["services"], ";") {
-			r.scrapeDiscoveredNode(n, now, errs)
-		}
-	}
-
-	return r.mb.Emit(), errs.Combine()
+	r.client = client
+	return nil
 }
 
-// scrapeNode collects metrics from a single Aerospike node
-func (r *aerospikeReceiver) scrapeNode(client aerospike, now pcommon.Timestamp, errs *scrapererror.ScrapeErrors) {
-	info, err := client.Info()
-	if err != nil {
-		errs.AddPartial(0, err)
-		return
+func (r *aerospikeReceiver) shutdown(_ context.Context) error {
+	r.logger.Debug("executing close")
+	if r.client != nil {
+		r.client.Close()
 	}
-
-	r.emitNode(info, now, errs)
-	r.scrapeNamespaces(info, client, now, errs)
+	return nil
 }
 
-// scrapeDiscoveredNode connects to a discovered Aerospike node and scrapes it using that connection
-//
-// If unable to parse the endpoint or connect, that error is logged and we return early
-func (r *aerospikeReceiver) scrapeDiscoveredNode(endpoint string, now pcommon.Timestamp, errs *scrapererror.ScrapeErrors) {
-	host, portStr, err := net.SplitHostPort(endpoint)
-	if err != nil {
-		r.logger.Warn(fmt.Sprintf("%s: %s", errBadEndpoint, err))
-		errs.Add(err)
-		return
-	}
-	port, err := strconv.ParseInt(portStr, 10, 32)
-	if err != nil {
-		r.logger.Warn(fmt.Sprintf("%s: %s", errBadPort, err))
-		errs.Add(err)
-		return
+// scrape scrapes both Node and Namespace metrics from the provided Aerospike node.
+// If CollectClusterMetrics is true, it then scrapes every discovered node
+func (r *aerospikeReceiver) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	r.logger.Sugar().Debug("beginning scrape")
+	errs := &scrapererror.ScrapeErrors{}
+
+	if r.client == nil {
+		var err error
+		r.logger.Sugar().Debug("client is nil, attempting to create a new client")
+		r.client, err = r.clientFactory(r.host, r.port)
+		if err != nil {
+			r.client = nil
+			addPartialIfError(errs, fmt.Errorf("client creation failed: %w", err))
+			return r.mb.Emit(), errs.Combine()
+		}
 	}
 
-	nClient, err := r.clientFactory(host, int(port))
-	if err != nil {
-		r.logger.Warn(err.Error())
-		errs.Add(err)
-		return
+	now := pcommon.NewTimestampFromTime(time.Now().UTC())
+	client := r.client
+
+	info := client.Info()
+	for _, nodeInfo := range info {
+		r.emitNode(nodeInfo, now, errs)
 	}
-	defer nClient.Close()
+	r.scrapeNamespaces(client, now, errs)
 
-	r.scrapeNode(nClient, now, errs)
+	return r.mb.Emit(), errs.Combine()
 }
 
 // emitNode records node metrics and emits the resource. If statistics are missing in INFO, nothing is recorded
 func (r *aerospikeReceiver) emitNode(info map[string]string, now pcommon.Timestamp, errs *scrapererror.ScrapeErrors) {
+	r.logger.Sugar().Debugf("emitNode len(info): %v", len(info))
```

Comment:
We seem to be sugaring the logger each time we use it. Should we do this once when constructing the component?

Reply:
changed to sugar once